test/test_device_histogram.cu

```diff
@@ -426,77 +430,107 @@ struct SearchTransform
     }
 };
 
-
 // Scales samples to evenly-spaced bins
 template <typename LevelT>
 struct ScaleTransform
 {
-    int    num_levels;  // Number of levels in array
-    LevelT max;         // Max sample level (exclusive)
-    LevelT min;         // Min sample level (inclusive)
-    LevelT scale;       // Bin scaling factor
-
-    void Init(
-        int    num_levels_,  // Number of levels in array
-        LevelT max_,         // Max sample level (exclusive)
-        LevelT min_,         // Min sample level (inclusive)
-        LevelT scale_)       // Bin scaling factor
-    {
-        this->num_levels = num_levels_;
-        this->max = max_;
-        this->min = min_;
-        this->scale = scale_;
-    }
+  int num_levels; // Number of levels in array
+  LevelT max;     // Max sample level (exclusive)
+  LevelT min;     // Min sample level (inclusive)
 
-    // Functor for converting samples to bin-ids  (num_levels is returned if sample is out of range)
-    template <typename SampleT>
-    int operator()(SampleT sample)
-    {
-        if ((sample < min) || (sample >= max))
-        {
-            // Sample out of range
-            return num_levels;
-        }
+  void Init(int num_levels_, // Number of levels in array
+            LevelT max_,     // Max sample level (exclusive)
+            LevelT min_)     // Min sample level (inclusive)
+  {
+    this->num_levels = num_levels_;
+    this->max        = max_;
+    this->min        = min_;
+  }
 
-        return (int) ((((LevelT) sample) - min) / scale);
+  // Functor for converting samples to bin-ids  (num_levels is returned if sample is out of range)
+  template <typename SampleT>
+  int operator()(SampleT sample)
+  {
+    if ((sample < min) || (sample >= max))
+    {
+      // Sample out of range
+      return num_levels;
     }
+
+    // Accurate bin computation following the arithmetic we guarantee in the HistoEven docs
+    return static_cast<int>(
+      (static_cast<uint64_t>(sample - min) * static_cast<uint64_t>(num_levels - 1)) /
```

Comment:
I don't think the initial [suggestion](https://github.com/NVIDIA/cub/pull/632/files#r1115268567) is addressed. If you'd like to repurpose the primary template for integral types only, please, add a static assert that `T` is integral. Also, there might be hard-to-find issues when we test histogram with `__{u,}int_t`, please, add a static assert for T size as well.